enhancements/siteconfig-extra-manifest-version-independence.md

```diff
@@ -0,0 +1,363 @@
+---
+title: Allowing version independence in siteconfig
+authors:
+  - "@sabbir-47"
+reviewers:
+  - "@imiller0"
+  - "@serngawy"
+  - "@Iack"
+  - "@serngawy" 
+approvers:
+  - "@imiller0"
+
+api-approvers:
+  - TBD
+creation-date: 2023-06-02
+last-updated: 2023-06-02
+status: provisional
+---
+
+
+# Cluster Group Upgrade (CGU) conditions and their reason/messages enhancement
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Documentation is updated in the upstream
+- [ ] Documentation is updated in the downstream
+
+
+## Summary
+
+Siteconfig allows customer to include custom extra manifests and list them in the siteconfig CR to be added during CR generation. It provides the flexibility to add additional CRs from the customer point of view during install time along with the CRs the RAN team provides. 
+
+The enhancement seek to address the version independence of extra manifests, by putting version dependent manifests in version specific folder in git and pointing the directory via **extraManifestPath** and disabling the fetching builtIn RAN CRs provided from the ztp container 
+
+
+## Motivation
+
+* Currently customer can put custom manifests in the git repo and point them via **extraManifestPath**. 
+* If they only want to include custom CRs to build site specific CRs, **inclusionDefault** must be set to *exclude*, and pass the list of file names under **filter.include**, but it does not allow to have the [same CR name](https://github.com/openshift-kni/cnf-features-deploy/blob/master/ztp/siteconfig-generator/siteConfig/siteConfigBuilder.go#L462-L464) for the custom CRs. 
+
+To be able to add custom CRs, customer needs to change the name of the custom CR.
+
+
+  ```yaml
+  - cluster:
+    extraManifestPath: mypath/
+    extraManifests:
+      filter:
+        inclusionDefault: exclude
+        include:
+          - myCR.yaml					
+  ```
+
+
+### Goals
+
+- Allow customer to include any custom CRs having same/different name than builtIn CRs residing in the container
+- Provide a better control on how to enable and disable builtIn CRs
+- Keep backward compatibility of the behavior corresponds to 4.13 version.
+
+### Non-Goals
+
+- TBD
+
+
+### User Stories
+
+- [EPIC LINK](https://issues.redhat.com/browse/CNF-7365) --> [User Story](https://issues.redhat.com/browse/CNF-8672)
+
+## Proposal
+
+* **Controllability:** The proposal is to introduce a parameter called  <span style="color:red">*extraManifests.builtinCRs*</span>. which can hold value "*include|exclude*", which provide the flexibility for the customer to include/exclude CRs that is provided via ztp-container. 
```

Comment:
Why do we need to list the files. Shouldn't it be a version directory which the site config consumes, if it overlaps with the default set, the new user defined CR takes precedence

Reply:
In the current version (4.13), 
- if it overlaps with the default set, precedence doesn't take place, and error is returned mentioning the filename must change. https://github.com/openshift-kni/cnf-features-deploy/blob/master/ztp/siteconfig-generator/siteConfig/siteConfigBuilder.go#L461-L465

Reply:
My point is that is should, not sure why we complicate this with adding lists, up to the user to put into the directory the CR's they want to add/override

Reply:
I like this idea.  If the user provides a duplicate filename, we should assume they intend to replace what was in the original.

To exclude a file, the user could create an empty file in their override directory with no contents, and we could take that as a signal to replace the objects in that file with nothing.

Reply:
For example...

Built-in directory tree:
```
source-crs
 - A.yaml (has object a1 and a2)
 - B.yaml (has objects b1 and b2)
 - C.yaml (has objects c1 and c2)
```
User provides:
```
 - A.yaml (has objects a3 and a4)
 - C.yaml (is empty)
```

The end result should be:
```
- a3 and a4 (because the user's A.yaml replaced the built-in A.yaml)
- b1 and b2 (because the user didn't override B.yaml)
- That is all (do NOT include objects c1 or c2 because the user provided an empty C.yaml)
```

Reply:
Currently, user can't say, i only want to apply my custom CRS and nothing should come from the ztp-container. The only workaround of that issue is suppose you have (A,B,C,D,E).yaml in the extraManifestPath and want to include all of them, then:

```yaml
extraManifestPath: mypath/
    extraManifests:
      filter:
        inclusionDefault: exclude
        exclude:
          - A.yaml
          - - A.yaml
          - - A.yaml
          - - A.yaml
```

Reply:
@lack  For the second case, if we support override and fallback as you have mentioned-

For the case of 2 different versions of extra-manifests, for example if a user have below directory structure 

```yaml
siteconfig/
├── kustomization.yaml
├── version_4.12
│   ├── kustomization.yaml
│   ├── sno-extra-manifest/
│   ├── cnfde9.yaml
└── version_4.13
    ├── kustomization.yaml
    ├── sno-extra-manifest/
    ├── cnfde10.yaml    
```

And the siteconfig points the extraManifestPath towards 4.12 folder 
```
 extraManifestPath: version_4.12/sno-extra-manifest
```
but ztp-container currently running is 4.13, falling back for one of the missing CR which may have changed from 4.12 to 4.13, will mismatch the version of the CR

Reply:
Doesn't it need to point to 4.13 if running 4.13 container ?

Reply:
"To exclude a file, the user could create an empty file in their override directory with no contents, and we could take that as a signal to replace the objects in that file with nothing." - What if there are 10 CR files they would like to exclude for example, now do they have to create 10 different empty files? I don't think its user friendly.

I think the reasonable solution for the case:

- User wants to add all the Custom CRs, none from the ztp container:

 ```yaml
  - cluster:
     extraManifestPath: mypath/
     extraManifests:
      builtinCRs: exclude
  ```
  
  - User wants to add Custom CRS along with the CRs from the container, but same named CR would take precedence from the user directory:
  
```extraManifestPath: mypath/
    extraManifests:
      builtinCRs: include	
```

Reply:
@browsell The idea is to decouple what the version we have in the container to what user wants to apply, if they have v4.10,v4.11,v4.12,v4.13 folders for example, and they have already extracted those CRs from the container and pushed to their git, then for deploying every version they will need to run same versioned ztp-container running each time in the hub

Reply:
Sounds complicated, you are going to have to build backwards compatibility into each new container version including different per release default crs.

Reply:
The goal of this work is to get to version independence (and not have to build in backwards compatibility to the tools/container). When we have the ability to exclude the built in content (this new `builtinCRs: exclude` field) the user can set their system up to only ever pull from git. This means they can have a set of 4.12, 4.13, 4.14, ... extra manifests in git and point to the correct set based on what they are deploying (one tool supporting multiple versions). 

Ideally we would set the default value of `builtinCRs` to "exclude" but that would break existing deployments in non-obvious ways. Existing customers, with no extra manifests in git would start deploying clusters with no extra manifests. Cluster would deploy just fine but would be seriously mis-configured. Perhaps we can flag the field for changing default in an upcoming release and communicate to customers the expectation that the extra manifests be kept in git alongside the SiteConfig starting with the current release?

Reply:
New commit 3dadf26972770907a83001b5feddd3236416f943 is added based on the design discussion happened in 06/09 (https://docs.google.com/document/d/1S4RVy6dRmOTuVtPngB8YM2bUp4CUED_370XZJKPgoPo/edit#). Please review @imiller0 @lack @serngawy @pixelsoccupied @browsell